File: mobile/packages/libghostty-terminal/ios/ExpoLibghosttyView.swift
import ExpoModulesCore
import GhosttyTerminal
import UIKit

private final class OrcaTerminalView: TerminalView {
  var keyboardEnabled = false
  var onCleanTap: ((CGPoint) -> Void)?

  // Why: Fabric sizes Expo views from intrinsicContentSize; a Metal grid
  // width here keeps the terminal at ~half the pane and wraps at ~35 cols.
  override var intrinsicContentSize: CGSize {
    CGSize(width: UIView.noIntrinsicMetric, height: UIView.noIntrinsicMetric)
  }

  override func toggleSoftwareKeyboard() {
    guard keyboardEnabled else { return }
    super.toggleSoftwareKeyboard()
  }

  @objc private func reportTap(_ recognizer: UITapGestureRecognizer) {
    onCleanTap?(recognizer.location(in: self))
  }

  func installTapReporter() {
    let recognizer = UITapGestureRecognizer(target: self, action: #selector(reportTap))
    recognizer.cancelsTouchesInView = false
    addGestureRecognizer(recognizer)
  }
}

// Ghostty-backed terminal view. The JS side owns transport and session
// lifecycle; this view only renders bytes pushed via `write` and reports
// user input / grid resizes back through events.
class ExpoLibghosttyView: ExpoView {
  private let terminalView = OrcaTerminalView(frame: .zero)
  private var session: InMemoryTerminalSession?
  private var viewportColumns = 0
  private var viewportRows = 0

  private func makeSession() -> InMemoryTerminalSession {
    InMemoryTerminalSession(
      write: { [weak self] data in
        self?.onInput([
          "data": data.base64EncodedString(),
          "text": String(decoding: data, as: UTF8.self),
        ])
      },
      resize: { [weak self] viewport in
        self?.viewportColumns = Int(viewport.columns)
        self?.viewportRows = Int(viewport.rows)
        self?.onResize(["cols": viewport.columns, "rows": viewport.rows])
      }
    )
  }

  private func installSession(_ session: InMemoryTerminalSession) {
    self.session = session
    terminalView.configuration = TerminalSurfaceOptions(
      backend: .inMemory(session),
      fontSize: fontSize
    )
  }

  /// Base font size in points (nil → ghostty's default, 14). Changing it on a
  /// mounted view rebuilds the surface, which resets the grid — set it before
  /// mounting. Pinch-to-zoom steps from this value.
  var fontSize: Float? {
    didSet {
      guard fontSize != oldValue, let session else { return }
      terminalView.configuration = TerminalSurfaceOptions(
        backend: .inMemory(session),
        fontSize: fontSize
      )
    }
  }

  var keyboardEnabled: Bool {
    get { terminalView.keyboardEnabled }
    set { terminalView.keyboardEnabled = newValue }
  }

  /// Hosts keep inactive tabs mounted. `opacity: 0` does not hide CAMetalLayer,
  /// so a second surface paints beside the active one. Stop drawing and hide.
  var surfaceVisible = true {
    didSet {
      guard surfaceVisible != oldValue else { return }
      terminalView.isHidden = !surfaceVisible
      terminalView.setSurfaceVisible(surfaceVisible)
    }
  }

  override var intrinsicContentSize: CGSize {
    CGSize(width: UIView.noIntrinsicMetric, height: UIView.noIntrinsicMetric)
  }

  /// Theme colors, applied through the shared controller's ghostty config —
  /// on iOS the theme is app-wide (every terminal view of the controller),
  /// unlike Android's per-view application. Colors accept ghostty config
  /// syntax; nil clears back to the defaults.
  func applyTheme(_ theme: TerminalThemeRecord?) {
    var config = TerminalConfiguration()
    if let theme {
      if let value = theme.background { config = config.background(value) }
      if let value = theme.foreground { config = config.foreground(value) }
      if let value = theme.cursorColor { config = config.cursorColor(value) }
      if let value = theme.selectionBackground { config = config.selectionBackground(value) }
      if let value = theme.selectionForeground { config = config.selectionForeground(value) }
      if let palette = theme.palette {
        for (index, color) in palette.enumerated() where index < 256 {
          if let color { config = config.palette(index, color: color) }
        }
      }
    }
    _ = TerminalController.shared.setTheme(TerminalTheme(light: config, dark: config))
  }

  let onInput = EventDispatcher()
  let onResize = EventDispatcher()
  let onBell = EventDispatcher()
  let onTitleChange = EventDispatcher()
  let onDirectoryChange = EventDispatcher()
  let onTap = EventDispatcher()

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    clipsToBounds = true

    let session = makeSession()
    // Without a controller the coordinator never builds a surface
    // ("surface rebuild skipped: missing controller").
    terminalView.controller = TerminalController.shared
    installSession(session)
    terminalView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
    // VoiceOver: one focusable element; direct interaction keeps touches
    // (taps, selection) flowing to the terminal instead of being swallowed.
    terminalView.isAccessibilityElement = true
    terminalView.accessibilityLabel = "Terminal"
    terminalView.accessibilityTraits = [.allowsDirectInteraction]
    terminalView.delegate = self
    terminalView.onCleanTap = { [weak self] point in self?.reportTap(at: point) }
    terminalView.installTapReporter()
    addSubview(terminalView)
  }

  override func layoutSubviews() {
    super.layoutSubviews()
    terminalView.frame = bounds
    if bounds.width > 0, bounds.height > 0 {
      terminalView.fitToSize()
    }
  }

  /// Feed PTY output (terminal.output on the wire) into the grid.
  func write(_ data: Data) {
    session?.receive(data)
  }

  /// Feed PTY output as UTF-8 text, for string-based wires.
  func write(_ text: String) {
    session?.receive(text)
  }

  func reset(with text: String) {
    var data = Data([0x1b, 0x63, 0x1b, 0x5b, 0x33, 0x4a])
    data.append(contentsOf: text.utf8)
    session?.receive(data)
  }

  func paste(_ text: String) {
    _ = terminalView.paste(text: text)
  }

  private func reportTap(at point: CGPoint) {
    guard
      viewportColumns > 0,
      viewportRows > 0,
      terminalView.bounds.width > 0,
      terminalView.bounds.height > 0,
      let text = session?.readViewportText()
    else {
      onTap([:])
      return
    }
    let row = min(
      viewportRows - 1,
      max(0, Int(point.y / terminalView.bounds.height * CGFloat(viewportRows)))
    )
    let column = min(
      viewportColumns - 1,
      max(0, Int(point.x / terminalView.bounds.width * CGFloat(viewportColumns)))
    )
    let lines = text.split(separator: "\n", omittingEmptySubsequences: false)
    guard row < lines.count else {
      onTap([:])
      return
    }
    onTap(["lineText": String(lines[row]), "column": column])
  }

  /// Mark the underlying PTY as exited (terminal.exit on the wire).
  func finish(exitCode: UInt32) {
    session?.finish(exitCode: exitCode, runtimeMilliseconds: 0)
  }
}

// Terminal effects (OSC escapes) surfaced as component events.
extension ExpoLibghosttyView: TerminalSurfaceBellDelegate, TerminalSurfaceTitleDelegate,
  TerminalSurfacePwdDelegate {
  func terminalDidRingBell() {
    onBell([:])
  }

  func terminalDidChangeTitle(_ title: String) {
    onTitleChange(["title": title])
  }

  func terminalDidChangeWorkingDirectory(_ path: String) {
    onDirectoryChange(["path": path])
  }
}

File: mobile/packages/libghostty-terminal/android/src/main/java/expo/modules/libghostty/ExpoLibghosttyView.kt
package expo.modules.libghostty

import android.content.Context
import android.util.Base64
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

// Ghostty-backed terminal view. The JS side owns transport and session
// lifecycle; this view only renders bytes pushed via `write` and reports
// user input / grid resizes back through events.
class ExpoLibghosttyView(context: Context, appContext: AppContext) :
  ExpoView(context, appContext) {
  // The accessory bar toggles visibility outside React's knowledge; without
  // this, RN never re-runs our LinearLayout pass and the bar stays invisible.
  override val shouldUseAndroidLayout = true

  private val onInput by EventDispatcher<Map<String, Any>>()
  private val onResize by EventDispatcher<Map<String, Any>>()
  private val onBell by EventDispatcher<Map<String, Any>>()
  private val onTitleChange by EventDispatcher<Map<String, Any>>()
  private val onDirectoryChange by EventDispatcher<Map<String, Any>>()
  private val onTap by EventDispatcher<Map<String, Any>>()

  private val terminal = GhosttyTerminalView(context).also { view ->
    view.onInputBytes = { bytes ->
      onInput(
        mapOf(
          "data" to Base64.encodeToString(bytes, Base64.NO_WRAP),
          "text" to String(bytes, Charsets.UTF_8)
        )
      )
    }
    view.onGridResize = { cols, rows ->
      onResize(mapOf("cols" to cols, "rows" to rows))
    }
    view.onBell = { onBell(emptyMap()) }
    view.onTitleChange = { title -> onTitleChange(mapOf("title" to title)) }
    view.onDirectoryChange = { path -> onDirectoryChange(mapOf("path" to path)) }
    view.onTap = { lineText, column ->
      onTap(mapOf("lineText" to lineText, "column" to column))
    }
  }

  private val accessoryBar = TerminalAccessoryBar(context).also { bar ->
    bar.onKey = { keyCode -> terminal.sendAccessoryKey(keyCode) }
    bar.onText = { text -> terminal.sendAccessoryText(text) }
    bar.onPaste = { terminal.requestPaste() }
    bar.visibility = GONE
  }

  init {
    orientation = VERTICAL
    terminal.accessoryBar = accessoryBar
    addView(terminal, LayoutParams(LayoutParams.MATCH_PARENT, 0, 1f))
    addView(accessoryBar, LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT))
    // The bar rides the soft keyboard: visible only while the IME is up.
    // React Native does not reliably dispatch insets down to nested native
    // views, so also re-check on every size change (the keyboard resizes us).
    ViewCompat.setOnApplyWindowInsetsListener(this) { _, insets ->
      syncAccessoryBar(insets)
      insets
    }
  }

  private fun syncAccessoryBar(insets: WindowInsetsCompat) {
    val imeVisible = insets.isVisible(WindowInsetsCompat.Type.ime())
    accessoryBar.visibility = if (imeVisible) VISIBLE else GONE
    // Edge-to-edge windows don't resize for the keyboard — it just covers us.
    // Pad our bottom by the covered amount so the bar sits above the IME and
    // the terminal shrinks (regaining the hidden rows). Hosts that already
    // handle the keyboard leave no overlap, so the padding stays zero.
    val imeBottom = insets.getInsets(WindowInsetsCompat.Type.ime()).bottom
    val location = IntArray(2)
    getLocationInWindow(location)
    val gapBelow = rootView.height - (location[1] + height)
    val overlap = if (imeVisible) (imeBottom - gapBelow).coerceAtLeast(0) else 0
    if (paddingBottom != overlap) {
      setPadding(0, 0, 0, overlap)
      requestLayout()
    }
  }

  override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
    super.onSizeChanged(w, h, oldw, oldh)
    val rootInsets = rootWindowInsets ?: return
    syncAccessoryBar(WindowInsetsCompat.toWindowInsetsCompat(rootInsets, this))
  }

  /** Base font size in dp; applied live (the grid reflows in place). */
  fun setFontSize(dp: Float) = terminal.setFontSize(dp)

  fun setKeyboardEnabled(enabled: Boolean) = terminal.setKeyboardEnabled(enabled)

  fun setSurfaceVisible(visible: Boolean) {
    visibility = if (visible) VISIBLE else INVISIBLE
    terminal.visibility = if (visible) VISIBLE else INVISIBLE
  }

  /** Apply theme colors; a null record clears back to the defaults. */
  internal fun setTheme(theme: TerminalThemeRecord?) = terminal.setTheme(
    foreground = theme?.foreground,
    background = theme?.background,
    cursorColor = theme?.cursorColor,
    selectionBackground = theme?.selectionBackground,
    selectionForeground = theme?.selectionForeground,
    palette = theme?.palette?.toTypedArray()
  )

  /** Feed PTY output (terminal.output on the wire) into the grid. */
  fun write(data: ByteArray) = terminal.write(data)

  fun reset(data: ByteArray) = terminal.reset(data)

  fun paste(text: String) = terminal.paste(text)

  /** Mark the underlying PTY as exited (terminal.exit on the wire). */
  fun finish(exitCode: Int) = terminal.finish(exitCode)

  internal fun destroy() = terminal.destroy()
}
